gk/fib.c

```diff
@@ -93,10 +150,18 @@ neigh_get_ether_cache_locked(struct neighbor_hash_table *neigh,
 		return eth_cache;
 	}
 
-	ret = rte_hash_add_key(neigh->hash_table, &addr->ip);
-	if (ret >= 0) {
-		eth_cache = &neigh->cache_tbl[ret];
-		fill_up_ether_cache_locked(eth_cache, addr, iface);
+	eth_cache = find_empty_ether_cache_entry(neigh);
+	if (eth_cache == NULL)
+		return NULL;
+
+	ret = fill_up_ether_cache_locked(eth_cache, addr, iface, lcore_id);
+	if (ret < 0) {
+		clear_ether_cache(eth_cache);
+		return NULL;
+	}
+
+	ret = rte_hash_add_key_data(neigh->hash_table, &addr->ip, eth_cache);
```

Comment:
If you pass the index of `eth_cache` here instead of the address of `eth_cache`, we wouldn't need to change the rest of the code as you did in `neigh_del_ether_cache()`.

The following example does exactly that:
http://dpdk.org/doc/api/examples_2ipsec-secgw_2ipsec-secgw_8c-example.html#a62

Please test if the cast `(void *)i` in the example is really needed.